chia/server/start_service.py

```diff
@@ -175,68 +189,97 @@ async def _connect_peers_task_handler(self) -> None:
+                            self.stop_requested.set,
+                            self.root_path,
+                            self.config,
+                            self._connect_to_daemon,
+                            max_request_body_size=self.max_request_body_size,
+                        )
+                yield
+            finally:
+                self._log.info(f"Stopping service {self._service_name} at port {self._advertised_port} ...")
 
-        self._connect_peers_task = asyncio.create_task(self._connect_peers_task_handler())
+                # start with UPnP, since this can take a while, we want it to happen
+                # in the background while shutting down everything else
+                for port in self._upnp_ports:
+                    self.upnp.release(port)
 
-        self._log.info(
-            f"Started {self._service_name} service on network_id: {self._network_id} "
-            f"at port {self._advertised_port}"
-        )
+                self._log.info("Cancelling reconnect task")
+                if self._connect_peers_task is not None:
+                    self._connect_peers_task.cancel()
+                self._log.info("Closing connections")
+                self._server.close_all()
 
-        if self._rpc_info:
-            rpc_api, rpc_port = self._rpc_info
-            self.rpc_server = await start_rpc_server(
-                rpc_api(self._node),
-                self.self_hostname,
-                self.daemon_port,
-                uint16(rpc_port),
-                self.stop,
-                self.root_path,
-                self.config,
-                self._connect_to_daemon,
-                max_request_body_size=self.max_request_body_size,
-            )
+                if self.rpc_server is not None:
+                    self._log.info("Closing RPC server")
+                    self.rpc_server.close()
 
-    async def run(self) -> None:
-        try:
-            with Lockfile.create(service_launch_lock_path(self.root_path, self._service_name), timeout=1):
-                try:
-                    await self.start()
-                except:  # noqa E722
-                    self.stop()
-                    raise
-                finally:
-                    await self.wait_closed()
-        except LockfileError as e:
-            self._log.error(f"{self._service_name}: already running")
-            raise ValueError(f"{self._service_name}: already running") from e
+                self._log.info("Waiting for socket to be closed (if opened)")
+
+                self._log.info("Waiting for ChiaServer to be closed")
+                await self._server.await_closed()
+
+                if self.rpc_server:
+                    self._log.info("Waiting for RPC server")
+                    await self.rpc_server.await_closed()
+                    self._log.info("Closed RPC server")
+
+        # this is a blocking call, waiting for the UPnP thread to exit
+        self.upnp.shutdown()
```

Comment:
```
# NOTE: avoid start=False, this is presently used for corner case setup type tests
```

`start` being `False` is in support of present test patterns and the behavior should go away and not be leveraged.  It allows the service to actually be started later but still have the context manager make sure it gets shut down.

Reply:
https://github.com/chia-network/chia-blockchain/commit/6575d14c32a97488d7aeebd202132bcd2f302481

Reply:
I think we could easily reduce the indentation as you say, but I don't think it would make the code actually more clear.  Hopefully we can just get around to getting rid of the `start` control entirely.